client/src/main/java/io/pravega/client/stream/notifications/notifier/ScaleEventNotifier.java

```diff
@@ -0,0 +1,100 @@
+import java.util.function.Supplier;
+
+import io.pravega.client.state.StateSynchronizer;
+import io.pravega.client.stream.impl.ReaderGroupState;
+import io.pravega.client.stream.notifications.Listener;
+import io.pravega.client.stream.notifications.NotificationSystem;
+import io.pravega.client.stream.notifications.events.ScaleEvent;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.Synchronized;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ScaleEventNotifier extends AbstractEventNotifier<ScaleEvent> {
+
+    private static final int UPDATE_INTERVAL_SECONDS = Integer.parseInt(
+            System.getProperty("pravega.client.scaleEvent.poll.interval.seconds", String.valueOf(120)));
+    private final Supplier<StateSynchronizer<ReaderGroupState>> synchronizerSupplier;
+    private final AtomicBoolean pollingStarted = new AtomicBoolean();
+    @GuardedBy("$lock")
+    private ScheduledFuture<?> future;
+    @GuardedBy("$lock")
+    private StateSynchronizer<ReaderGroupState> synchronizer;
+
+    public ScaleEventNotifier(final NotificationSystem notifySystem,
+                              final Supplier<StateSynchronizer<ReaderGroupState>> synchronizerSupplier,
+                              final ScheduledExecutorService executor) {
+        super(notifySystem, executor);
+        this.synchronizerSupplier = synchronizerSupplier;
+    }
+
+    @Override
+    @Synchronized
+    public void registerListener(final Listener<ScaleEvent> listener) {
+        notifySystem.addListeners(getType(), listener, this.executor);
+        //periodically fetch the scale
+        if (!pollingStarted.getAndSet(true)) { //schedule the  only once
+            synchronizer = synchronizerSupplier.get();
+            future = executor.scheduleAtFixedRate(this::checkAndTriggerScaleNotification, 0,
+                    UPDATE_INTERVAL_SECONDS, TimeUnit.SECONDS);
+        }
+    }
+
+    @Override
+    @Synchronized
+    public void unregisterListener(final Listener<ScaleEvent> listener) {
+        notifySystem.removeListener(getType(), listener);
+        if (!notifySystem.isListenerPresent(getType())) {
+            cancelScheduledTask();
+            synchronizer.close();
+        }
+    }
+
+    @Override
+    @Synchronized
+    public void unregisterListeners() {
+        this.notifySystem.removeListeners(getType());
+        cancelScheduledTask();
+        synchronizer.close();
+    }
+
+    @Override
+    public String getType() {
+        return ScaleEvent.class.getSimpleName();
+    }
+
+    private void checkAndTriggerScaleNotification() {
+        this.synchronizer.fetchUpdates();
+        ReaderGroupState state = this.synchronizer.getState();
+        ScaleEvent event = ScaleEvent.builder().numOfSegments(state.getNumberOfSegments())
+                         .numOfReaders(state.getOnlineReaders().size())
+                         .build();
+        if (event.getNumOfReaders() != event.getNumOfSegments()) {
```

Comment:
This is limiting, we want to notify stream scaling even if the number of readers and segments is the same. It is up to the application to determine whether it is important that they are the same or not.

Reply:
Yes, it makes sense. Updated the condition to ensure notifications are generated only if there is atleast one reader. Plz review...